### PR TITLE
IGNITE-8640 earlier validation of cache config

### PR DESCRIPTION
**Pull request by dgarus**

### CODE REVIEW

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/CacheAffinitySharedManager.java [hunk 1]
```diff
@@ -2785,4 +2775,4 @@ private void saveCacheConfiguration(CacheConfiguration<?, ?> cfg, boolean sql) {
             }
         }
     }
-}
+}
```
Comment (SomeFire) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/cache/CacheAffinitySharedManager.java]: Need empty line at the end of file.

File: modules/core/src/test/java/org/apache/ignite/internal/processors/cache/IgniteAbstractDynamicCacheStartFailTest.java [hunk 2]
```diff
@@ -534,6 +624,29 @@ protected void testDynamicCacheStart(final Collection<CacheConfiguration> cfgs,
         }
     }
 
+    /** {@inheritDoc} */
+    @Override protected IgniteConfiguration getConfiguration(String igniteInstanceName) throws Exception {
+        IgniteConfiguration res = super.getConfiguration(igniteInstanceName);
+
+        if(mbSrv == null)
```
Comment (SomeFire) [inline modules/core/src/test/java/org/apache/ignite/internal/processors/cache/IgniteAbstractDynamicCacheStartFailTest.java]: Need space after `if`.

File: modules/core/src/test/java/org/apache/ignite/internal/processors/cache/IgniteAbstractDynamicCacheStartFailTest.java [hunk 3]
```diff
@@ -534,6 +624,29 @@ protected void testDynamicCacheStart(final Collection<CacheConfiguration> cfgs,
...
+        return res;
+    }
+
+    /** {@inheritDoc} */
+    @Override protected void afterTest() throws Exception {
+        super.afterTest();
+
+        mbSrv.clear();
+
+        for (String cacheName : grid(0).cacheNames())
```
Comment (SomeFire) [inline modules/core/src/test/java/org/apache/ignite/internal/processors/cache/IgniteAbstractDynamicCacheStartFailTest.java]: Multiline `for` statement should be wrapped with braces.

File: modules/core/src/test/java/org/apache/ignite/internal/processors/cache/IgniteAbstractDynamicCacheStartFailTest.java [hunk 4]
```diff
@@ -630,6 +743,55 @@ protected CacheConfiguration createCacheConfiguration(String cacheName) {
...
+     * @param useFilter {@code true} if NodeFilter should be used.
+     *
+     * @return List of cache configurations.
+     */
+    protected List<CacheConfiguration> createCacheConfigsWithFailureMbServer(
+        boolean failOnAllNodes,
+        int unluckyNode,
+        int unluckyCfg,
+        int cacheNum,
+        boolean useFilter
```
Comment (SomeFire) [inline modules/core/src/test/java/org/apache/ignite/internal/processors/cache/IgniteAbstractDynamicCacheStartFailTest.java]: `useFilter` is always false, may be remove it?

File: modules/core/src/test/java/org/apache/ignite/internal/processors/cache/IgniteAbstractDynamicCacheStartFailTest.java [hunk 5]
```diff
@@ -630,6 +743,55 @@ protected CacheConfiguration createCacheConfiguration(String cacheName) {
...
+
+            if (useFilter)
+                cfg.setNodeFilter(new NodeFilter(uuid));
+
+            cfgs.add(cfg);
+        }
+
+        mbSrv.node(getTestIgniteInstanceName(unluckyNode));
+
+        if (failOnAllNodes)
```
Comment (SomeFire) [inline modules/core/src/test/java/org/apache/ignite/internal/processors/cache/IgniteAbstractDynamicCacheStartFailTest.java]: Multiline `if` statement should be wrapped with braces. But I think we should move this check to the previous `for` and make check `i == unluckyCfg || failOnAllNodes` with appropriate actions in true-branch.

File: modules/core/src/test/java/org/apache/ignite/internal/processors/cache/IgniteAbstractDynamicCacheStartFailTest.java [hunk 6]
```diff
@@ -772,4 +934,254 @@ public BrokenStoreFactory(boolean eOnAllNodes, String gridName) {
                 return null;
         }
     }
+
+    /** Failure MBean server*/
+    private class FailureMBeanServer implements MBeanServer {
+
```
Comment (SomeFire) [inline modules/core/src/test/java/org/apache/ignite/internal/processors/cache/IgniteAbstractDynamicCacheStartFailTest.java]: Remove empty line after class definition.

File: modules/core/src/test/java/org/apache/ignite/internal/processors/cache/IgniteAbstractDynamicCacheStartFailTest.java [hunk 7]
```diff
@@ -772,4 +934,254 @@ public BrokenStoreFactory(boolean eOnAllNodes, String gridName) {
                 return null;
         }
     }
+
+    /** Failure MBean server*/
```
Comment (SomeFire) [inline modules/core/src/test/java/org/apache/ignite/internal/processors/cache/IgniteAbstractDynamicCacheStartFailTest.java]: Fix to `/** Failure MBean server. */`.

File: modules/core/src/test/java/org/apache/ignite/internal/processors/cache/IgniteAbstractDynamicCacheStartFailTest.java [hunk 8]
```diff
@@ -772,4 +934,254 @@ public BrokenStoreFactory(boolean eOnAllNodes, String gridName) {
...
+        }
+
+        /** Clear failure set of caches and set of nodes. */
+        void clear() {
+            caches.clear();
+            nodes.clear();
+        }
+
+        /** {@inheritDoc} */
+        @Override public ObjectInstance registerMBean(Object object, ObjectName name)
```
Comment (SomeFire) [inline modules/core/src/test/java/org/apache/ignite/internal/processors/cache/IgniteAbstractDynamicCacheStartFailTest.java]: Use abbreviations in variable names. `object` -> `obj`, `class` -> `cls`, `loader` -> `ldr`, `query` -> `qry`.

File: modules/core/src/test/java/org/apache/ignite/internal/processors/cache/IgniteAbstractDynamicCacheStartFailTest.java [hunk 9]
```diff
@@ -772,4 +934,254 @@ public BrokenStoreFactory(boolean eOnAllNodes, String gridName) {
...
+        }
+
+        /** {@inheritDoc} */
+        @Override public ObjectInstance getObjectInstance(ObjectName name) throws InstanceNotFoundException {
+            return origin.getObjectInstance(name);
+        }
+
+        /** {@inheritDoc} */
+        @Override public Set<ObjectInstance> queryMBeans(ObjectName name,
+            QueryExp query) {
```
Comment (SomeFire) [inline modules/core/src/test/java/org/apache/ignite/internal/processors/cache/IgniteAbstractDynamicCacheStartFailTest.java]: Second line isn't needed for this method.

File: modules/core/src/test/java/org/apache/ignite/internal/processors/cache/IgniteAbstractDynamicCacheStartFailTest.java [hunk 10]
```diff
@@ -772,4 +934,254 @@ public BrokenStoreFactory(boolean eOnAllNodes, String gridName) {
...
+            return origin.queryMBeans(name, query);
+        }
+
+        /** {@inheritDoc} */
+        @Override public Set<ObjectName> queryNames(ObjectName name,
```
Comment (SomeFire) [inline modules/core/src/test/java/org/apache/ignite/internal/processors/cache/IgniteAbstractDynamicCacheStartFailTest.java]: Second line isn't needed for this method.

File: modules/core/src/test/java/org/apache/ignite/internal/processors/cache/IgniteAbstractDynamicCacheStartFailTest.java [hunk 11]
```diff
@@ -772,4 +934,254 @@ public BrokenStoreFactory(boolean eOnAllNodes, String gridName) {
...
+        }
+
+        /** {@inheritDoc} */
+        @Override public Integer getMBeanCount() {
+            return origin.getMBeanCount();
+        }
+
+        /** {@inheritDoc} */
+        @Override public Object getAttribute(ObjectName name,
+            String attribute) throws MBeanException, AttributeNotFoundException, InstanceNotFoundException, ReflectionException {
```
Comment (SomeFire) [inline modules/core/src/test/java/org/apache/ignite/internal/processors/cache/IgniteAbstractDynamicCacheStartFailTest.java]: Too long line. Move second argument to the previous line.

File: modules/core/src/test/java/org/apache/ignite/internal/processors/cache/IgniteAbstractDynamicCacheStartFailTest.java (hunk 8 above)
Comment (dgarus) [reply modules/core/src/test/java/org/apache/ignite/internal/processors/cache/IgniteAbstractDynamicCacheStartFailTest.java]: done

File: modules/core/src/test/java/org/apache/ignite/internal/processors/cache/IgniteAbstractDynamicCacheStartFailTest.java (hunk 10 above)
Comment (dgarus) [reply modules/core/src/test/java/org/apache/ignite/internal/processors/cache/IgniteAbstractDynamicCacheStartFailTest.java]: done

File: modules/core/src/test/java/org/apache/ignite/internal/processors/cache/IgniteAbstractDynamicCacheStartFailTest.java (hunk 11 above)
Comment (dgarus) [reply modules/core/src/test/java/org/apache/ignite/internal/processors/cache/IgniteAbstractDynamicCacheStartFailTest.java]: done

File: modules/core/src/test/java/org/apache/ignite/internal/processors/cache/IgniteAbstractDynamicCacheStartFailTest.java (hunk 7 above)
Comment (dgarus) [reply modules/core/src/test/java/org/apache/ignite/internal/processors/cache/IgniteAbstractDynamicCacheStartFailTest.java]: done

**Pull request closed without merging**